### PR TITLE
Linear Algebra Primer

### PR DESCRIPTION
**Pull request by ben-albrecht**
**LinearAlgebra** contents:

 - Compiling
 - Basics
   - Linear Algebra Types
   - Array operations
  - Factory Functions
    - Vectors
    - Matrices
  - Operations
  - Properties
  - Structure
  - Caveats
  - LinearAlgebra.Sparse
    - Supported Sparse Layouts
    - Factory Functions
    - Operations

## Testing

- [x] `test/modules/packages/LinearAlgebra/correctness`
- [x] `test/release/examples/primers/LinearAlgebralib.chpl`

### CODE REVIEW

File: test/release/examples/primers/LinearAlgebralib.chpl [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+  Basics
+  ------
+
+  Linear Algebra Types
+  ~~~~~~~~~~~~~~~~~~~~
+
+  In Chapel, there are no ``Vector`` or ``Matrix`` types.
+  Instead, vectors and matrices are represented as 1D and 2D native Chapel
+  arrays, respectively. As a result, the terms *vector* and *1D array* are used
+  interchangably, as well as *matrix* and *2D array*.
```
Comment (cassella) [inline test/release/examples/primers/LinearAlgebralib.chpl]: interchangeably

Comment (ben-albrecht) [reply test/release/examples/primers/LinearAlgebralib.chpl]: Thanks **@chplspell**

File: test/release/examples/primers/LinearAlgebralib.chpl [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+  // Prints:
+  // 0.0 0.0 0.0 0.0 0.0
+var v1 = Vector(5);       // from a length
+var v2 = Vector(0..#5);   // from a range
+var v3 = Vector({0..#5}); // from a domain
+var v4 = Vector([0, 0, 0, 0, 0], eltType=real); // from a 1D array
+var v5 = Vector(0, 0, 0, 0, 0, eltType=real);   // from a list of values as arguments
+
+// If the ``eltType`` is omitted for the argument case, it will be inferred from
+// the first argument, as shown here:
```
Comment (benharsh) [inline test/release/examples/primers/LinearAlgebralib.chpl]: optional: maybe make it clear that the result is an array of reals?

File: test/release/examples/primers/LinearAlgebralib.chpl [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+var AtB = A.times(B);
+assert(AtB.rank == 2); // not flattened!
+
+// Element-wise division (avoiding promotion flattening)
+var AdB = A.elementDiv(B);
+assert(AdB.rank == 2); // not flattened!
+
+// Taking the transpose of a matrix:
+transpose(M0);
+M0.T; // short-hand syntax
```
Comment (benharsh) [inline test/release/examples/primers/LinearAlgebralib.chpl]: This kind of looks like ``transpose`` and ``T`` might modify ``M0`` in-place.

File: test/release/examples/primers/LinearAlgebralib.chpl [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+{
+  // Bad example of LinearAlgebra.Sparse usage:
+  var A = CSRMatrix({1..100, 1..100}); // Create a CSR matrix from dense domain
+  writeln(A.domain); // empty sparse domain, with no way to add elements..
+}
+
+/*
+  The above sparse matrix is initialized with an empty domain, and the user
+  has no way of accessing that domain. Therefore, it is not possible to add
+  indices to the sparse domain
```
Comment (benharsh) [inline test/release/examples/primers/LinearAlgebralib.chpl]: Missing period.

File: test/release/examples/primers/LinearAlgebralib.chpl [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+var D3 = CSRDomain(0..#3);             // from 1 range
+var D4 = CSRDomain(0..#3, 0..#3);      // from 2 ranges
+var D5 = CSRDomain({0..#3, 0..#3});    // from a 2D domain
+
+// Creating CSR matrices:
+var M1 = CSRMatrix(D1);             // from a CSR domain
+var M2 = CSRMatrix({0..#3, 0..#3}); // from a dense domain
+var M3 = CSRMatrix(M1);             // From a CSR matrix
+const I = eye(3,3);
+var M4 = CSRMatrix(I);              // From a dense matrix
```
Comment (benharsh) [inline test/release/examples/primers/LinearAlgebralib.chpl]: Is it worth stating here that only the indices corresponding to non-zero values are added to the sparse domain, or is it reasonable to expect users to look at the full docs? Up to you.

File: test/release/examples/primers/LinearAlgebralib.chpl [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+a.size; // 15
+
+// Shape
+a.shape; // (3, 5)
+
+// Element type
+writeln(a.eltType: string); // real(64)
+
+// Array type in format of:
+//   ``[domain(dimensions, index-type, stridable)] element-type``
```
Comment (benharsh) [inline test/release/examples/primers/LinearAlgebralib.chpl]: dimensions --> rank

**Pull request merged**